language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/LSPackageCache.java

```diff
@@ -49,60 +53,52 @@ public LSPackageCache() {
      * @return {@link BLangPackage} BLang Package resolved
      */
     public BLangPackage findPackage(CompilerContext compilerContext, PackageID pkgId) {
-        if (containsPackage(pkgId.getName().getValue())) {
-            return packageMap.get(pkgId.getName().getValue());
+        BLangPackage bLangPackage = packageCache.get(pkgId);
+        if (bLangPackage != null) {
+            return bLangPackage;
         } else {
-            BLangPackage bLangPackage = LSPackageLoader.getPackageById(compilerContext, pkgId);
+            bLangPackage = LSPackageLoader.getPackageById(compilerContext, pkgId);
             addPackage(bLangPackage);
             return bLangPackage;
         }
     }
 
-    /**
-     * get the package name by composing from given identifier list.
-     *
-     * @param pkgID             package ID
-     * @return {@link String}   Full package name with orgName
-     */
-    private String getPackageName(PackageID pkgID) {
-        return pkgID.getOrgName().getValue() + "/" + pkgID.getName().getValue();
-    }
-
     /**
      * add package to the package map.
      *
      * @param bLangPackage ballerina package to be added.
      */
     void addPackage(BLangPackage bLangPackage) {
         if (bLangPackage.getPackageDeclaration() == null) {
-            this.packageMap.put(".", bLangPackage);
+            //TODO check whether getPackageDeclaration() is needed
+            this.packageCache.put(new PackageID(Names.DOT.value), bLangPackage);
         } else {
-            this.packageMap.put(getPackageName(bLangPackage.packageID), bLangPackage);
+            this.packageCache.put(bLangPackage.packageID, bLangPackage);
         }
     }
-
-    /**
-     * Check whether the package exist in the current package context.
-     * @param packageName       package name
-     * @return {@link Boolean}  package exist or not
-     */
-    private boolean containsPackage(String packageName) {
-        return this.packageMap.containsKey(packageName);
-    }
     
-    private void loadPackagesMap() {
-        CompilerContext tempCompilerContext = CommonUtil.prepareTempCompilerContext();
+    private void loadPackagesMap(CompilerContext tempCompilerContext) {
         for (String staticPkgName : staticPkgNames) {
             PackageID packageID = new PackageID(new org.wso2.ballerinalang.compiler.util.Name("ballerina"),
                     new org.wso2.ballerinalang.compiler.util.Name(staticPkgName),
                     new org.wso2.ballerinalang.compiler.util.Name("0.0.0"));
             
-            this.packageMap.put(getPackageName(packageID),
-                    LSPackageLoader.getPackageById(tempCompilerContext, packageID));
+            this.packageCache.put(packageID, LSPackageLoader.getPackageById(tempCompilerContext, packageID));
```

Comment:
As a secondary improvement, we can check here whether the package has been already added.

Reply:
@nadeeshaan `LSPackageLoader.getPackageById()` seems internally updates the cache. Thus no need to update the cache here. Calling only `LSPackageLoader.getPackageById()` would be sufficient.